alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java

```diff
@@ -714,39 +698,37 @@ public void finished(final Environment<Double, Euclidean2DPosition> env, final T
             }
 
             private boolean thereIsOverlap(final Environment<Double, Euclidean2DPosition> env) {
-                final boolean posResult =  env.getNodes().stream()
+                return !env.getNodes().stream()
                         .filter(n -> n instanceof CellWithCircularArea)
-                        .map(n -> env.getNodesWithinRange(n, ((CellWithCircularArea<Euclidean2DPosition>) n).getDiameter()).stream()
-                                .filter(c -> env.getDistanceBetweenNodes(c, n) < (((CellWithCircularArea<Euclidean2DPosition>) n).getDiameter() * MAX_PRECISION))
-                                .collect(Collectors.toList()).isEmpty())
-                        .allMatch(b -> b);
-                if (posResult) {
-                    return posResult;
-                } else {
-                    env.getNodes().stream()
-                    .filter(n -> n instanceof CellWithCircularArea)
-                    .forEach(n -> {
-                        final List<Node<Double>> listOverlapping = env.getNodesWithinRange(n, (((CellWithCircularArea) n).getDiameter())).stream()
-                                .filter(c -> env.getDistanceBetweenNodes(c, n) < ((CellWithCircularArea) n).getDiameter())
-                                .collect(Collectors.toList());
-                        if (!listOverlapping.isEmpty()) {
-                            System.out.println("nodes: ");
-                            System.out.println("center : " + env.getPosition(n));
-                            listOverlapping.forEach(c -> System.out.println("In range : " + env.getPosition(c)));
-                            listOverlapping.forEach(c -> System.out.println("distance : " + env.getPosition(c).getDistanceTo(env.getPosition(n))));
-                        }
-                    });
-                    return posResult;
-                }
+                        .map(n -> (CellWithCircularArea<Euclidean2DPosition>) n)
+                        .flatMap(n -> env.getNodesWithinRange(n, n.getDiameter() - DELTA).stream().peek(c -> {
```

Comment:
Peeking on a stream is very, very non functional. Try to collect the information, and fail only on termination:
```java
n -> env.getNodesWithinRange(n, n.getDiameter() - DELTA).stream()
    .map(n -> new Pair(c, n))
```